mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,516 @@
+use std::any::Any;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use crate::{Event, EventStreamer};
+use common::executor::abortable_queue::WeakSpawner;
+use common::log;
+
+use futures::channel::mpsc::UnboundedSender;
+use futures::channel::oneshot;
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use tokio::sync::mpsc;
+
+/// The errors that could originate from the streaming manager.
+#[derive(Debug)]
+pub enum StreamingManagerError {
+    /// There is no streamer with the given ID.
+    StreamerNotFound,
+    /// Couldn't send the data to the streamer.
+    SendError(String),
+    /// The streamer doesn't accept an input.
+    NoDataIn,
+    /// Couldn't spawn the streamer.
+    SpawnError(String),
+    /// The client is not known/registered.
+    UnknownClient,
+    /// A client with the same ID already exists.
+    ClientExists,
+    /// The client is already listening to the streamer.
+    ClientAlreadyListening,
+}
+
+struct StreamerInfo {
+    /// The communication channel to the streamer.
+    data_in: Option<UnboundedSender<Box<dyn Any + Send>>>,
+    /// Clients the streamer is serving for.
+    clients: HashSet<u64>,
+    /// The shutdown handle of the streamer.
+    shutdown: oneshot::Sender<()>,
+}
+
+impl StreamerInfo {
+    fn new(data_in: Option<UnboundedSender<Box<dyn Any + Send>>>, shutdown: oneshot::Sender<()>) -> Self {
+        Self {
+            data_in,
+            clients: HashSet::new(),
+            shutdown,
+        }
+    }
+
+    fn add_client(&mut self, client_id: u64) { self.clients.insert(client_id); }
+
+    fn remove_client(&mut self, client_id: &u64) { self.clients.remove(client_id); }
+
+    fn is_down(&self) -> bool { self.shutdown.is_canceled() }
+}
+
+struct ClientInfo {
```

Comment:
I wonder about the whole client infrastructure. It reminds me the pub/sub logic and maybe too complicated (client_id needs to be generated by GUI and is appeared in so many structs).
Maybe I am missing something, but could we not implement this with the standard channel api?
Like an enable rpc starts a streamer, if needed, and returns its streamer_id. An RpcTaskManager instance also returns its streamer_id. Then GUI connects to the SSE endpoint with the streamer_id and SSE_handler retrieves the channel rx part from the streamer and listens for events.
I think a channel like watch can handle well it when one or several or no clients are connected.

(I guess, maybe reason for client_id is that it allows GUI to get events from all streamers in one SSE connection. But could we not just pass several streamer_ids in the same url query for that and listen on multiple receivers?)

Reply:
> Like an enable rpc starts a streamer, if needed, and returns its streamer_id

then the client has to perform one more connection to actually subscribe (a new SSE connection), that's per streamer.

> An RpcTaskManager instance also returns its streamer_id

well, task managed rpc endpoints are actually streamer-less (they push events but has no underlying streamer), we could def make a streamer for them but that would be an extra thread per task request.

> Then GUI connects to the SSE endpoint with the streamer_id and SSE_handler retrieves the channel rx part from the streamer and listens for events.
I think a channel like watch can handle well it when one or several or no clients are connected.

i think the problem with this approach is now that the server has to have one backward SSE connection per client per streamer. Assuming a single client that's still probably too many backward connections when we could have multiplexed everything into a single one.

> (I guess, maybe reason for client_id is that it allows GUI to get events from all streamers in one SSE connection. But could we not just pass several streamer_ids in the same url query for that and listen on multiple receivers?)

this is actually the approach that was in use before (we called them filters), but the problem with this is that the client will need to terminate their SSE connection and re-open it again if they want to unsubscribe or subscribe to a single streamer. So imagine if a client has 10 tx_history streamers which they are subscribed to and init-ed a time consuming task (e.g. `task::account_balance::init`) now they would need to disconnect their SSE connection and reconnect again with the same 10 tx_history streamers plus the new task streamer. When the task is over they would need to disconnect and reconnect again with the 10 streamers without the task one.
Well we really have double the number of connections & disconnections with each new streamer added compared to the case where we don't multiplex streamers together.

Reply:
> well, task managed rpc endpoints are actually streamer-less (they push events but has no underlying streamer)

I think the task manager could be itself be a streamer. When an task is started via rpc it would return its streamer_id

> client will need to terminate their SSE connection and re-open it again if they want to unsubscribe or subscribe to a single streamer.

Thank you for the explanation. I see the problem now. Maybe indeed it is convenient for GUI to have a single permanent connection and just add or remove new event listeners.

But maybe we still could remove client_ids from the existing structs and make a 'broker' object which would redistribute events from streamers to subscribers. I guess we would need an rpc to subscribe/unsubscribe to a task rpc streamer. 
(I understand that we need client_id to generate it on GUI side so we could continue listening to the same events if we reconnect.)
It looks like event has event type field. Maybe it is convenient for GUI if it uses same event type both in rpc to enable the streamer and on the page to listen to events.

Reply:
covered here [91d98ed](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2172/commits/91d98edbefc3edd1b68c8bbf912c2db31a5718ed)